src/Compilers/Core/AnalyzerDriver/AnalyzerManager.cs

```diff
@@ -0,0 +1,138 @@
+                }, cancellationToken));
+        }
+
+        private Task<HostSessionStartAnalysisScope> GetSessionAnalysisScopeAsync(
+            DiagnosticAnalyzer analyzer,
+            Action<Diagnostic> addDiagnostic,
+            Func<Exception, DiagnosticAnalyzer, bool> continueOnAnalyzerException,
+            CancellationToken cancellationToken)
+        {
+            Func<DiagnosticAnalyzer, Task<HostSessionStartAnalysisScope>> getTask = a =>
+            {
+                return Task.Run(() =>
+                {
+                    var sessionScope = new HostSessionStartAnalysisScope();
+                    AnalyzerDriverHelper.ExecuteInitializeMethod(a, sessionScope, addDiagnostic, continueOnAnalyzerException, cancellationToken);
+                    return sessionScope;
+                }, cancellationToken);
+            };
+
+            var task = ImmutableInterlocked.GetOrAdd(ref _sessionScopeMap, analyzer, getTask);
+
+            // Retry cancelled task.
+            if (task.Status == TaskStatus.Canceled)
+            {
+                ImmutableInterlocked.TryUpdate(ref _sessionScopeMap, analyzer, getTask(analyzer), task);
+                return _sessionScopeMap[analyzer];
+            }
+
+            return task;
+        }
+
+        /// <summary>
+        /// Get all the analyzer actions to execute for the given analyzer against a given compilation.
+        /// The returned actions include the actions registered during <see cref="DiagnosticAnalyzer.Initialize(AnalysisContext)"/> method as well as
+        /// the actions registered during <see cref="CompilationStartAnalyzerAction"/> for the given compilation.
+        /// </summary>
+        public async Task<AnalyzerActions> GetAnalyzerActionsAsync(
+            DiagnosticAnalyzer analyzer,
+            Compilation compilation,
+            Action<Diagnostic> addDiagnostic,
+            AnalyzerOptions analyzerOptions,
+            Func<Exception, DiagnosticAnalyzer, bool> continueOnAnalyzerException,
+            CancellationToken cancellationToken)
+        {
+            var sessionScope = await GetSessionAnalysisScopeAsync(analyzer, addDiagnostic, continueOnAnalyzerException, cancellationToken).ConfigureAwait(false);
+            if (sessionScope.CompilationStartActions.Length > 0 && compilation != null)
+            {
+                var compilationScope = await GetCompilationAnalysisScopeAsync(analyzer, sessionScope,
+                    compilation, addDiagnostic, analyzerOptions, continueOnAnalyzerException, cancellationToken).ConfigureAwait(false);
+                return compilationScope.GetAnalyzerActions(analyzer);
+            }
+
+            return sessionScope.GetAnalyzerActions(analyzer);
+        }
+
+        /// <summary>
+        /// Return <see cref="DiagnosticAnalyzer.SupportedDiagnostics"/> of given <paramref name="analyzer"/>.
+        /// </summary>
+        public ImmutableArray<DiagnosticDescriptor> GetSupportedDiagnosticDescriptors(
```

Comment:
for this to be here, its lifetime must be same as host (in VS case, VS session)

otherwise, cache can't do its expected enforcement. (calling GetSupportedDiagnostics only once per analyzer)

if its expected lifetime is same as host, fields can't hold onto diagnostic analyzer. more specifically project diagnostic analyzer. otherwise we have memory leak. is this checked in?